Expose vLLM Metrics to serve.llm API (#52719)
ray-project/ray: doc/source/serve/llm/serving-llms.rst

Ray Serve config file, `serve_TIMESTAMP.yaml`, that you can reference and re-run in the
future.

After reading and reviewing the generated model config, refer to
the vLLM engine configuration `docs <https://docs.vllm.ai/en/latest/serving/engine_args.html>`_
for further customization.

Enable logging
---------------------
Ray enables LLM service-level logging by default, and makes these statistics available using Grafana and Prometheus. Instructions are available at :ref:`collect-metrics`.

These higher-level metrics track request and token behavior across deployed models. For example, with Prometheus metric names inline:

* Token Volumes: `ray_rayllm_tokens_input`

  * Total input and generated tokens (last hour, 24 hours, and 7 days)
  * Ratio of input to generated tokens

* Tokens per Request: `ray_rayllm_requests_started`

  * Average total tokens per request (overall and per model)

> * Peak and Aggregate Stats: `ray_rayllm_tokens_generated`

Reviewer: Can we find and replace rayllm with serve_llm across all of this PR?

Author: Done

Author: <img width="1840" alt="Screenshot 2025-05-12 at 5 09 19 PM" src="https://github.com/user-attachments/assets/3ac09501-aea5-420b-a142-df99175e27dd" />
Tested dashboard still works

  * Peak tokens per second (per model, per day)
  * Per-model request and token gauges over the past week

For visualization, Ray ships with a Serve LLM-specific dashboard which is automatically available in Grafana. Example below:

.. image:: images/serve_llm_dashboard.png

If enabled, all engine metrics (this currently implies vLLM) are available through the Ray metrics export endpoint and are queryable using Prometheus. See `vLLM metrics <https://docs.vllm.ai/en/stable/serving/metrics.html>`_ for a complete list. These are also visualized by the Serve LLM Grafana dashboard. For example:

> * Token Throughput: `ray_vllm:request_prompt_tokens_sum` `ray_vllm:generation_tokens_total`)=

Reviewer: the equal sign does not make sense in the end.

                ),
                deployment_config=dict(
                    autoscaling_config=dict(
                        min_replicas=1, max_replicas=2,
                    )
                ),
                log_engine_metrics=True
            )

            app = build_openai_app({"llm_configs": [llm_config]})
            serve.run(app, blocking=True)

    .. tab-item:: YAML
        :sync: bind

        .. code-block:: yaml

            # config.yaml
            applications:
            - args:
                llm_configs:
                    log_engine_metrics: true
                    - model_loading_config:
                        model_id: qwen-0.5b
                        model_source: Qwen/Qwen2.5-0.5B-Instruct
                    accelerator_type: A10G
                    deployment_config:
                        autoscaling_config:
                            min_replicas: 1
                            max_replicas: 2
                    accelerator_type: A10G
                    deployment_config:
                        autoscaling_config:
                            min_replicas: 1
                            max_replicas: 2

>             import_path: ray.serve.llm:build_openai_app

Reviewer: The yaml is not right. log_engine_metrics should be in the list separator.

Ray Serve config file, `serve_TIMESTAMP.yaml`, that you can reference and re-run in the
future.

> After reading and reviewing the generated model config, refer to

Reviewer 2: ```suggestion
After reading and reviewing the generated model config, see
```

the `vLLM engine configuration docs <https://docs.vllm.ai/en/latest/serving/engine_args.html>`_
for further customization.

Observability
---------------------
Ray enables LLM service-level logging by default, and makes these statistics available using Grafana and Prometheus. For more details on configuring Grafana and Prometheus, see :ref:`collect-metrics`.

These higher-level metrics track request and token behavior across deployed models. For example: average total tokens per request, ratio of input tokens to generated tokens, and peak tokens per second.

> For visualization, Ray ships with a Serve LLM-specific dashboard which is automatically available in Grafana. Example below:

Reviewer 2: ```suggestion
For visualization, Ray ships with a Serve LLM-specific dashboard, which is automatically available in Grafana. Example below:
```

Ray Serve config file, `serve_TIMESTAMP.yaml`, that you can reference and re-run in the
future.

After reading and reviewing the generated model config, refer to
the vLLM engine configuration `docs <https://docs.vllm.ai/en/latest/serving/engine_args.html>`_
for further customization.

> Enable logging

Author: "Enable logging" -> "Observability". Lmk thoughts

> Enable logging

Reviewer 3: ```suggestion
Enable metrics logging
```

---------------------
Ray enables LLM service-level logging by default, and makes these statistics available using Grafana and Prometheus. Instructions are available at :ref:`collect-metrics`.

> These higher-level metrics track request and token behavior across deployed models. For example, with Prometheus metric names inline:

Author: Makes sense. Going to shift these inline to make clear they're examples and add full docs with [LLM-1894](https://anyscale1.atlassian.net/browse/LLM-1894?atlOrigin=eyJpIjoiZTg3ZWFjODI2MGNmNDAxNWI4MjNlY2Q2ZTQ4OTI0Y2IiLCJwIjoiaiJ9)

* Token Volumes: `ray_serve_llm_tokens_input`

  * Total input and generated tokens (last hour, 24 hours, and 7 days)
  * Ratio of input to generated tokens

* Tokens per Request: `ray_serve_llm_requests_started`

  * Average total tokens per request (overall and per model)

* Peak and Aggregate Stats: `ray_serve_llm_tokens_generated`

  * Peak tokens per second (per model, per day)
  * Per-model request and token gauges over the past week

For visualization, Ray ships with a Serve LLM-specific dashboard which is automatically available in Grafana. Example below:

.. image:: images/serve_llm_dashboard.png

> If enabled, all engine metrics (this currently implies vLLM) are available through the Ray metrics export endpoint and are queryable using Prometheus. See `vLLM metrics <https://docs.vllm.ai/en/stable/serving/metrics.html>`_ for a complete list. These are also visualized by the Serve LLM Grafana dashboard. For example:

Author: New section -> "Engine Metrics"
